gcovr/json_generator.py

```diff
@@ -31,21 +50,65 @@ def print_json_report(covdata, output_file, options):
 
     for no in sorted(covdata):
         gcovr_json_file = {}
-        gcovr_json_file['file'] = presentable_filename(covdata[no].filename, root_filter=options.root_filter)
+        gcovr_json_file['file'] = presentable_filename(covdata[no].filename,
+                                                       root_filter=options.root_filter)
         gcovr_json_file['lines'] = _json_from_lines(covdata[no].lines)
         gcovr_json_root['files'].append(gcovr_json_file)
 
-    write_json = json.dump
-    if options.prettyjson:
-        write_json = functools.partial(write_json, indent=PRETTY_JSON_INDENT, separators=(',', ': '), sort_keys=True)
-    else:
-        write_json = functools.partial(write_json, sort_keys=True)
+    _write_json_result(gcovr_json_root, output_file, options.prettyjson)
 
-    if output_file is not None:
-        with open(output_file, 'w') as output:
-            write_json(gcovr_json_root, output)
-    else:
-        write_json(gcovr_json_root, sys.stdout)
+
+#
+# Produce gcovr JSON summary report
+#
+def print_json_summary_report(covdata, output_file, options):
+
+    total_lines = 0
+    total_covered = 0
+
+    json_dict = {}
+    json_dict['current_working_directory'] = options.root
+    json_dict['gcovr/summary_format_version'] = JSON_SUMMARY_FORMAT_VERSION
+    json_dict['files'] = []
+
+    # Data
+    keys = sort_coverage(
+        covdata, show_branch=options.show_branch,
+        by_num_uncovered=options.sort_uncovered,
+        by_percent_uncovered=options.sort_percent)
+
+    def _summarize_file_coverage(coverage):
+        filename = presentable_filename(coverage.filename,
+                                        root_filter=options.root_filter)
+
+        if options.show_branch:
+            total, cover, percent = coverage.branch_coverage()
+            uncovered_lines = coverage.uncovered_branches_str()
+        else:
+            total, cover, percent = coverage.line_coverage()
+            uncovered_lines = coverage.uncovered_lines_str()
```

Comment:
Instead of switching between line and branch coverage, the summary can reasonably show both. Compare #376, as they essentially show the same data.

Reply:
I understand the idea, and the data is essentially right there; however, currently, the text/tabulated summary does not show branch. Are we departing from that? What should be the behavior of the summary with and without show_branch option?

Reply:
The text report switches between line/branch coverage for space reasons, not because this is fundamentally useful. For a machine-readable report, text width does not matter. Including this data is effectively zero cost but makes the summary much more useful.

It may be useful to use the same field names as the CSV report (#376) uses for columns.